### PR TITLE
`Development`: Unify admin endpoints

### PR DESCRIPTION
**Pull request by julian-christl**
#### General

- [X] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [X] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [X] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [X] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
#### Client
- [X] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.

### Motivation and Context

The idea of these changes is to prepare the endpoints for automated authorization tests. Additionally, it allows for autocompletion.

### Description

I moved all endpoints that are only accessible to Admins from `api/...` to `api/admin/...`, wrapped `@PreAuthorize("hasRole('ADMIN')")` in `@EnforceAdmin` and replaced all occurrences.
I also adapted the documentation and the pull request template.

### Steps for Testing

Look at the code and consider checking all affected endpoints (I also listed some functionality below). If there is no difference between endpoints in regard to the changes, testing them might not be necessary.
Keep in mind that this only affects permissions, not functionality. It's enough to check if you can see a resource/execute some action successfully.

Admin functionality:
- Create and delete courses
- Everything in the top right menu for admins
    - Upcoming exams & exercises
    - Manage users
    - Manage organisations
    - etc
- Trigger automatic assessment for text exercises

### Steps for Code Review
I made sure that all occurrences were replaced. Please double-check that I didn't accidentally replace too much and changed the permissions.
Also, ensure that I adapted all connections and didn't damage the interfaces.
If you have suggestions on how to improve the separation, let me know.

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### CODE REVIEW

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]
Left some remarks (see comments). Two more things:
- Could you also add a guideline to the documentation on when to add a new AdminResource instead of adding the admin endpoint to the same resource?
- You should be able to put @EnforceAdmin over the class declaration in the Admin Resources, since all endpoints in there should only have admin privileges

File: src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java:144]: Such things exist throughout the system. Fixing this is out of scope. However, these changes I do lead potentially to some way of detecting such things in the future quite easily.

Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java:144]: For now I override the ID to be sure but in the future we hopefully find an improvement for this.

File: src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java:194]: Nope. In the Client we only create through the POST endpoint. I checked that. Otherwise, I would've changed that..

File: src/main/java/de/tum/in/www1/artemis/web/rest/ModelingExerciseResource.java
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ModelingExerciseResource.java:330]: I now decided to move all to dedicated classes. It's probably better this way. This also resolves your first question from above.
All admin stuff is in one package.

**Comment by julian-christl** [comment]
> * You should be able to put @EnforceAdmin over the class declaration in the Admin Resources, since all endpoints in there should only have admin privileges

This is possible, yes. I actually decided against that now and adapted the code.
The intention is that a developer can see the required role without the need to scroll up.
This also prevents overrides of the annotation.

**Review by JohannesStoehr — DISMISSED** [review]
Thanks for incorporating the feedback. Code and documentation look good to me

File: src/main/java/de/tum/in/www1/artemis/security/annotations/EnforceAdmin.java
Comment (stefanwaldhauser) [inline src/main/java/de/tum/in/www1/artemis/security/annotations/EnforceAdmin.java:20]: 💡Idea: This is a great idea actually. Maybe we could create this for other roles also? Without changing every occurence in this PR of course

File: src/main/java/de/tum/in/www1/artemis/web/rest/admin/AdminUserResource.java
Comment (stefanwaldhauser) [inline src/main/java/de/tum/in/www1/artemis/web/rest/admin/AdminUserResource.java:47]: ❌ Required Change: The logger seems to refer to the wrong class. Please check the other occurrences also

File: src/main/java/de/tum/in/www1/artemis/web/rest/admin/AdminOrganizationResource.java
Comment (stefanwaldhauser) [inline src/main/java/de/tum/in/www1/artemis/web/rest/admin/AdminOrganizationResource.java:33]: ❌ Required Change: The logger seems to refer to the wrong class. Please check the other occurrences also

File: src/main/java/de/tum/in/www1/artemis/web/rest/admin/AdminStatisticsResource.java
Comment (stefanwaldhauser) [inline src/main/java/de/tum/in/www1/artemis/web/rest/admin/AdminStatisticsResource.java:27]: ❌ Required Change: The logger seems to refer to the wrong class. Please check the other occurrences also

File: src/main/java/de/tum/in/www1/artemis/web/rest/admin/AdminSystemNotificationResource.java
Comment (stefanwaldhauser) [inline src/main/java/de/tum/in/www1/artemis/web/rest/admin/AdminSystemNotificationResource.java:28]: ❌ Required Change: The logger seems to refer to the wrong class. Please check the other occurrences also

File: src/main/java/de/tum/in/www1/artemis/web/rest/admin/ManagementResource.java
Comment (stefanwaldhauser) [inline src/main/java/de/tum/in/www1/artemis/web/rest/admin/ManagementResource.java:16]: ❓ Question: Is there are reason why this class has no Admin Prefix?

File: src/main/java/de/tum/in/www1/artemis/web/rest/admin/AuditResource.java
Comment (stefanwaldhauser) [inline src/main/java/de/tum/in/www1/artemis/web/rest/admin/AuditResource.java:28]: ❓ Question: Is there are reason why this class has no Admin Prefix?

File: src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java
Comment (stefanwaldhauser) [inline src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java:197]: ❓ Question: This feels a little bit strange. Should the updated course not have an ID already?

**Review by stefanwaldhauser — CHANGES_REQUESTED** [review]
Good Job! Some requested changes ❌, ideas 💡 and questions ❓. Only changes ❌ need to be addressed for approval.

File: src/main/java/de/tum/in/www1/artemis/security/annotations/EnforceAdmin.java
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/security/annotations/EnforceAdmin.java:20]: That is the plan. But it's one PR at a time.

File: src/main/java/de/tum/in/www1/artemis/web/rest/admin/ManagementResource.java
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/web/rest/admin/ManagementResource.java:16]: Yes, see JavaDoc in @EnforceAdmin

File: src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java:197]: The issue here is that the course could have another ID. An attacker could then update a course they don't have access to.

As fixing the endpoint is out of scope, I decided to override the ID to ensure that the correct course gets updated.

**Review by stefanwaldhauser — DISMISSED** [review]
Reapprove

**Review by b-fein — COMMENTED** [review]
I’ve only got a few small suggestions for the updated documentation and two for consistency of method names.

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]
Seams like some e2e tests are failing

**Comment by julian-christl** [comment]
> Seams like some e2e tests are failing

I don't think this is an issue with my code. If you look on Bamboo, the last build that ran through was the merge commit.
I got two commits after that:
1. Only touches two tests
2. Touches documentation, automatic renaming of two methods and only one change in the system notifications service that first should not change the functionality and second should not affect the tests that are failing now.

Considering that a rerun of [this](https://bamboo.ase.in.tum.de/chain/result/viewChainResult.action?planKey=ARTEMIS-AETG888&buildNumber=8) (based on the same develop commit) yields the same result, I would say this is a general issue right now.

**Review by stefanwaldhauser — DISMISSED** [review]
Reapprove

**Review by b-fein — APPROVED** [review]
Code: :+1:

**Comment by julian-christl** [comment]
> > > Seams like some e2e tests are failing
> > 
> > 
> > I don't think this is an issue with my code. If you look on Bamboo, the last build that ran through was the merge commit. I got two commits after that:
> > 
> > 1. Only touches two tests
> > 2. Touches documentation, automatic renaming of two methods and only one change in the system notifications service that first should not change the functionality and second should not affect the tests that are failing now.
> > 
> > Considering that a rerun of [this](https://bamboo.ase.in.tum.de/chain/result/viewChainResult.action?planKey=ARTEMIS-AETG888&buildNumber=8) (based on the same develop commit) yields the same result, I would say this is a general issue right now.
> 
> They are consistently failing in your PR, even if I rerun the tests. But they do pass on develop, so you might want to take a look at them...

They do not pass on develop. The latest one (your PR) is also failing. However, it passed before the merge.
See the discussion on Slack in #artemisdev we currently have. It looks like the Bamboo prelive is at fault.

**Review by JohannesStoehr — APPROVED** [review]
In that case...
Code looks good to me

**Review by bassner — DISMISSED** [review]
Code looks good.
Gave it a test on TS5, too - admin endpoints seem to work correctly.
I tried accessing some of the admin functionality using a hacked client that *thinks* it's an admin: Admins endpoints consistently returned 403.
LGTM 👍

**Review by b-fein — DISMISSED** [review]
code re-approve

**Comment by julian-christl** [comment]
Got a bit premature with the first fix but now it's working.
Found here: https://github.com/ls1intum/Artemis/pull/5638#pullrequestreview-1117750106

Tested on TS1

**Review by b-fein — DISMISSED** [review]
code re-approve

**Review by JohannesStoehr — APPROVED** [review]
Code looks good after latest merges and path fix

**Review by stefanwaldhauser — DISMISSED** [review]
Good change! LGTM!

**Comment by julian-christl** [comment]
I cherry-picked some commits, and others I reimplemented. So some new code reviews to check for consistency would make sense

File: src/main/java/de/tum/in/www1/artemis/service/user/UserService.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/user/UserService.java:4]:
```suggestion
import static de.tum.in.www1.artemis.config.Constants.*;
```

**Review by JohannesStoehr — DISMISSED** [review]
Why do you have that duplicated import?

File: src/main/java/de/tum/in/www1/artemis/service/user/UserService.java
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/service/user/UserService.java]: Weird combination of Auto-forrmatting and spotless probably

**Review by JohannesStoehr — DISMISSED** [review]
Looked though the most important code changes. Code looks good to me

File: docs/dev/guidelines/server.rst
Comment (Strohgelaender) [inline docs/dev/guidelines/server.rst:309]: You are currently ignoring the return value of this call.
```suggestion
        authCheckService.checkHasAtLeastRoleForExerciseElseThrow(Role.INSTRUCTOR, exercise, null);
```

File: docs/dev/guidelines/server.rst
Comment (Strohgelaender) [inline docs/dev/guidelines/server.rst:326]: All other roles are in caps but Editor is not, is this correct?

Comment (julian-christl) [reply docs/dev/guidelines/server.rst:326]:
```suggestion
| EDITOR           | @PreAuthorize("hasRole('EDITOR')")     |
```

File: src/main/java/de/tum/in/www1/artemis/service/user/UserService.java
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/service/user/UserService.java:509]: Single responsibility: I suggest extracting this block and the other one below to two separate methods `checkUsernameValidityElseThrow()` and `checkPasswordValidityElseThrow()`.

File: src/main/java/de/tum/in/www1/artemis/web/rest/admin/AdminCourseResource.java
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/web/rest/admin/AdminCourseResource.java:124]:
```suggestion

```
`elseThrow` above, so `course` cannot be `null`

File: src/main/java/de/tum/in/www1/artemis/web/rest/admin/AdminExamResource.java
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/web/rest/admin/AdminExamResource.java:49]: Those should not be needed (and not have been needed before) as the annotation already ensures that, right?

```suggestion
```

Most of the admin endpoints do not check the authorization again, so I think we should remove it consistently.

File: src/main/java/de/tum/in/www1/artemis/web/rest/admin/AdminExerciseResource.java
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/web/rest/admin/AdminExerciseResource.java:44]:
```suggestion
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/admin/AuditResource.java
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/web/rest/admin/AuditResource.java:43]:
```suggestion
    @GetMapping("audits")
    @EnforceAdmin
```
Not really important, but in all other resources the annotation order is swapped.

**Review by b-fein — COMMENTED** [review]
Code changes look good. Only left some minor comments about small inconsistencies.

Some comments may also be about copied code that could be improved with a small fix. Feel free to ignore those.

Manual testing will follow later today.

File: src/main/java/de/tum/in/www1/artemis/web/rest/admin/AdminExamResource.java
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/web/rest/admin/AdminExamResource.java:33]: now unused import

```suggestion

/**
 * REST controller for administrating Exam.
 */
@RestController
@RequestMapping("api/admin/")
public class AdminExamResource {

    private final Logger log = LoggerFactory.getLogger(AdminExamResource.class);

    private final ExamRepository examRepository;

    public AdminExamResource(ExamRepository examRepository) {
        this.examRepository = examRepository;
    }
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/admin/AdminExerciseResource.java
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/web/rest/admin/AdminExerciseResource.java:26]: unused

File: src/main/java/de/tum/in/www1/artemis/web/rest/admin/AuditResource.java
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/web/rest/admin/AuditResource.java:59]:
```suggestion
    @GetMapping(value = ("audits"), params = { "fromDate", "toDate" })
    @EnforceAdmin
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/admin/AuditResource.java
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/web/rest/admin/AuditResource.java:78]:
```suggestion
    @GetMapping("audits/{id:.+}")
    @EnforceAdmin
```

**Review by laadvo — DISMISSED** [review]
Tested on ts1, everything worked as expected

**Review by b-fein — APPROVED** [review]
Code re-approve.

Manual testing: All CRUD operations for courses, system notifications, users, organisations, and modelling clusters still work. All the other pages in the server administration menu also show the intended information. :+1:

**Comment by julian-christl** [comment]
In #5723, we fundamentally change how authentication works. Therefore, we won't have a dedicated file authentication anymore, so I reverted the file endpoint commit.

**Review by b-fein — APPROVED** [review]
Re-approve code. Changes do not affect admin endpoints. Therefore, manual testing review still applies.

**Review by laadvo — APPROVED** [review]
Everything still working, can reapprove. Tested on ts1

**Review by JohannesStoehr — APPROVED** [review]
Re-approve code

**Pull request merged**
